modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheAffinitySharedManager.java

```diff
@@ -1982,7 +1982,37 @@ public boolean onCentralizedAffinityChange(final GridDhtPartitionsExchangeFuture
         });
 
         synchronized (mux) {
-            this.waitInfo = null;
+            waitInfo = null;
+        }
+
+        return true;
+    }
+
+    /**
+     * Called on exchange initiated by baseline server node leave.
+     *
+     * @param fut Exchange future.
+     * @param crd Coordinator flag.
+     * @return {@code True} if affinity should be assigned by coordinator.
+     */
+    public boolean onBaselineNodeLeft(final GridDhtPartitionsExchangeFuture fut, boolean crd) {
+        assert (fut.events().hasServerLeft() && !fut.firstEvent().eventNode().isClient()) : fut.firstEvent();
+
+        forAllRegisteredCacheGroups(new IgniteInClosureX<CacheGroupDescriptor>() {
+            @Override public void applyx(CacheGroupDescriptor desc) throws IgniteCheckedException {
+                CacheGroupHolder cache = getOrCreateGroupHolder(fut.initialVersion(), desc);
+
+                cache.aff.reinitializeWithoutOfflineNodes(fut.initialVersion());
```

Comment:
This is not correct affinity initialization.
You don't take into account that cluster may have `MOVING` partitions. In this case, a primary node can be chosen from `MOVING` backup and affinity is not recalculated after PME end. You can see correct affinity recalculation at `org.apache.ignite.internal.processors.cache.CacheAffinitySharedManager#onServerLeftWithExchangeMergeProtocol`

Reply:
Since I'm using assignment (not idealAssignment), which took moving into account (correct?) this should be еру correct initialization?

Reply:
There is no guarantee that the next backup will have a partition in OWNING status.
Case 1:
Suppose that you have a partition with 2 backups. Primary is OWNER other nodes are MOVING:
OMM
The second backup completed rebalancing and now have OWNING partition state:
OMO
Now the primary node is left and you have:
MO in affinity distribution. MOVING partition becomes primary which is incorrect behavior.
Case 2:
You lost your last OWNING partition in affinity. In this case, detectLostPartitions should be triggered.

You also nullify WaitRebalanceInfo. You shouldn't do this because in this case if you have currently running rebalancing late affinity assignment will not be triggered after rebalancing completion.

Reply:
1) Let's clarify, aff.assignment contains only owning partitions (correct?)
I'm just shrinking previous assignment, without recalculation.
I see no MOVING issues here, am I missed something?

2) Rebalance will be restarted in case of moving partitions.
See no problem here :(

Reply:
1. Affinity assignment contains OWNING and MOVING partitions.
2. Rebalance is started but late affinity assignment because of nullified WaitRebalanceInfo - not.
3. Detect lost partitions should be triggered after PME because if you don't have OWNING partitions you should mark others as LOST due to loss policy.